### PR TITLE
Use cursors with iterator in `aggregate` command

### LINKED ISSUE #1733: Support `getMore` command
### What should be done?

https://www.mongodb.com/docs/manual/reference/command/getMore/

- [x] Do not consume all documents in `find`: #2933
- [x] Do not consume all documents in `aggregate`
- [ ] Refactor tests (https://github.com/FerretDB/FerretDB/pull/2929/files#r1246365875)

Test script: #2761

### LINKED ISSUE #2877: Do not consume all documents in `processStagesDocuments`
### What should be done?

`processStagesDocuments` should return an iterator that should be stored in a cursor, not a slice of docs.

TBD

### Where?

TBD

### Definition of Done

- all handlers updated;
- unit tests added/updated;
- integration/compatibility tests added/updated;
- spot refactorings done;
- user documentation updated or an issue to create documentation created;
- something else?

### PR DESCRIPTION
**Pull request by chilagrow**
# Description

Closes #2877.
Closes #1733.

## Readiness checklist

- [ ] I added/updated unit tests.
- [x] I added/updated integration/compatibility tests.
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: integration/aggregate_documents_compat_test.go [hunk 1]
```diff
@@ -92,6 +95,12 @@ func testAggregateStagesCompatWithProviders(t *testing.T, providers shareddata.P
 				pipeline = append(pipeline, bson.D{{"$sort", bson.D{{"_id", 1}}}})
 			}
 
+			opts := new(options.AggregateOptions)
```
Comment (AlekSi) [inline integration/aggregate_documents_compat_test.go:98]: I think we should use `options.Aggregate()` like we do with `options.FindOne()`, `options.Update()`, `options.Delete()`, etc. Currently, there are two outliers: `options.AggregateOptions` and `options.IndexOptions`

File: integration/aggregate_documents_compat_test.go [hunk 2]
```diff
@@ -92,6 +95,12 @@ func testAggregateStagesCompatWithProviders(t *testing.T, providers shareddata.P
...
+
+			if tc.maxTime != nil {
+				opts.SetMaxTime(pointer.GetDuration(tc.maxTime))
```
Comment (AlekSi) [inline integration/aggregate_documents_compat_test.go:101]:
```suggestion
				opts.SetMaxTime(*tc.maxTime)
```

We already know it is not nil

File: internal/handlers/pg/msg_aggregate.go [hunk 3]
```diff
@@ -86,6 +86,17 @@ func (h *Handler) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
...
 
+	v, _ := document.Get("maxTimeMS")
+	if v == nil {
+		v = int64(0)
+	}
+
+	maxTimeMS, err := commonparams.GetValidatedNumberParamWithMinValue(document.Command(), "maxTimeMS", v, 0)
+	if err != nil {
+		// unreachable for MongoDB GO driver, it validates maxTimeMS parameter
+		return nil, err
```
Comment (AlekSi) [inline internal/handlers/pg/msg_aggregate.go:97]: lazyerror

File: internal/handlers/pg/msg_aggregate.go [hunk 4]
```diff
@@ -86,6 +86,17 @@ func (h *Handler) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
...
 	username, _ := conninfo.Get(ctx).Auth()
 
+	v, _ := document.Get("maxTimeMS")
+	if v == nil {
+		v = int64(0)
+	}
+
+	maxTimeMS, err := commonparams.GetValidatedNumberParamWithMinValue(document.Command(), "maxTimeMS", v, 0)
+	if err != nil {
+		// unreachable for MongoDB GO driver, it validates maxTimeMS parameter
```
Comment (AlekSi) [inline internal/handlers/pg/msg_aggregate.go:96]: I'm not sure if I understand that comment correctly, but if it says that this code is not reachable by tests, then this is already visible from them, and we have to way to update that comment when/if this line becomes reachable.

File: internal/handlers/pg/msg_aggregate.go [hunk 5]
```diff
@@ -166,7 +177,16 @@ func (h *Handler) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
 		return nil, err
 	}
 
-	var resDocs []*types.Document
+	cancel := func() {}
+	if maxTimeMS != 0 {
+		// It is not clear if maxTimeMS affects only find, or both find and getMore (as the current code does).
```
Comment (AlekSi) [inline internal/handlers/pg/msg_aggregate.go:182]: It is `aggregate` handler, not `find`

File: internal/handlers/pg/msg_aggregate.go [hunk 6]
```diff
@@ -253,32 +275,40 @@ type stagesDocumentsParams struct {
...
+
+	if err := p.dbPool.InTransactionKeep(ctx, func(tx pgx.Tx) error {
+		keepTx = tx
 
-	if err := p.dbPool.InTransaction(ctx, func(tx pgx.Tx) error {
 		var err error
-		iter, _, err := pgdb.QueryDocuments(ctx, tx, p.qp)
+		iter, _, err = pgdb.QueryDocuments(ctx, tx, p.qp)
 		if err != nil {
 			return err
```
Comment (AlekSi) [inline internal/handlers/pg/msg_aggregate.go:286]: lazyerrors

**Review by AlekSi — COMMENTED** [review]
LGTM overall, but I wonder if we have tests for a combination of `aggregate`+`getMore` and, if not, if we could add them without copy&pasting `find`+`getMore` tests

File: internal/util/iterator/values.go [hunk 7]
```diff
@@ -85,17 +85,21 @@ func ConsumeValuesN[K, V any](iter Interface[K, V], n int) ([]V, error) {
 }
 
 // Values returns an iterator over values of another iterator.
+// It will be added to the given closer.
 //
 // Close method closes the underlying iterator.
 // For that reason, there is no need to track both iterators.
-func Values[K, V any](iter Interface[K, V]) Interface[struct{}, V] {
+func Values[K, V any](iter Interface[K, V], closer *MultiCloser) Interface[struct{}, V] {
```
Comment (AlekSi) [inline internal/util/iterator/values.go:92]: Oh, no, please revert that. This function always closes the iterator.

Comment (AlekSi) [inline internal/util/iterator/values.go]: Let's revert that for now. Let's keep MultiCloser for handlers only

Comment (chilagrow) [reply internal/util/iterator/values.go]: Okie, I've reverted. I noticed that not all stages handled closer upon creating iterator and I thought adding that would make it easier.

Comment (AlekSi) [reply internal/util/iterator/values.go]: I'm not completely sure about that yet, but I think we will change the behavior of iterator's Close method working "back" towards the begging on the chain by closing the underlying iterator. If we do that, then iterator's Close and MultiCloser will be clearly separated, and we could add it to all iterators

File: internal/handlers/pg/msg_aggregate.go (hunk 4 above)
Comment (chilagrow) [reply internal/handlers/pg/msg_aggregate.go:96]: So values that are negative and not int64 return driver error and doesn't call our backend. However, we wouldn't know when driver may change the behavior, if that happens failing gracefully (possibly wrong error code and message) is best we can do.

Meanwhile I could add tests for just there is an error without checking error type or content.

**Comment by chilagrow** [comment]
> LGTM overall, but I wonder if we have tests for a combination of `aggregate`+`getMore` and, if not, if we could add them without copy&pasting `find`+`getMore` tests

@AlekSi, currently combination tests for `aggregate` and `getMore` tests batchSizes, but it's implicit because getMore is called via cursor https://github.com/FerretDB/FerretDB/blob/main/integration/aggregate_documents_test.go#L821 Do we think we need to expand?

`getMore` tests are mainly done with `find` at the moment like this one https://github.com/FerretDB/FerretDB/blob/main/integration/query_test.go#L959

File: integration/aggregate_documents_compat_test.go (hunk 1 above)
Comment (chilagrow) [reply integration/aggregate_documents_compat_test.go:98]: I will standadise options.AggregateOptions and options.IndexOptions to use options.Aggregate() and  options.Index()

File: internal/handlers/pg/msg_aggregate.go (hunk 4 above)
Comment (chilagrow) [reply internal/handlers/pg/msg_aggregate.go:96]: Honestly not sure adding this test was good idea. Other PRs seems to remove driver tests so... We could easily revert https://github.com/FerretDB/FerretDB/pull/2929/commits/0945325285532113967c6b74d10df7024fa08076 if we want to.

File: integration/query_test.go [hunk 8]
```diff
@@ -623,8 +624,8 @@ func TestQueryCommandBatchSize(t *testing.T) {
 
 	// the number of documents is set above the default batchSize of 101
 	// for testing unset batchSize returning default batchSize
-	docs := generateDocuments(0, 110)
```
Comment (chilagrow) [inline integration/query_test.go:626]: This file may be collapsed by github, please review it

File: integration/query_test.go [hunk 9]
```diff
@@ -1239,87 +1241,125 @@ func TestQueryCommandGetMore(t *testing.T) {
...
 
-			var res bson.D
-			err := collection.Database().RunCommand(ctx, findCommand).Decode(&res)
-			require.NoError(t, err)
+			for _, command := range []bson.D{findCommand, aggregateCommand} {
+				var res bson.D
+				err := collection.Database().RunCommand(ctx, command).Decode(&res)
+				require.NoError(t, err)
 
-			v, ok := res.Map()["cursor"]
```
Comment (chilagrow) [inline integration/query_test.go:1255]: A lot of change around here is making linter happy with recent deprecation of Map() https://github.com/FerretDB/FerretDB/actions/runs/5410541556/jobs/9832136007

**Review by w84thesun — APPROVED** [review]
👍

**Review by noisersup — APPROVED** [review]
Implementation looks good to me

File: internal/handlers/pg/msg_aggregate.go (hunk 4 above)
Comment (chilagrow) [reply internal/handlers/pg/msg_aggregate.go:96]: That commit link above tests `runCommand` already and it returns driver error :sweat_smile:

Comment (w84thesun) [reply internal/handlers/pg/msg_aggregate.go:96]: Oh, I missed that 🙈

**Review by AlekSi — APPROVED** [review]

**Pull request merged**